namui/sample/hooks-v2/src/hooks/v2/hooks/hooks/state.rs

```diff
@@ -0,0 +1,129 @@
+use super::*;
+
+pub fn use_state<'a, State: Send + Sync + Debug + 'static>(
+    init: impl FnOnce() -> State,
+) -> (Sig<'a, State>, SetState<State>) {
+    let ctx = ctx();
+
+    let instance = ctx.instance.as_ref();
+    let mut state_list = instance.state_list.lock().unwrap();
+
+    let state_index = ctx
+        .state_index
+        .fetch_add(1, std::sync::atomic::Ordering::SeqCst);
+
+    let no_state = || state_list.len() <= state_index;
+
+    if no_state() {
+        let state = init();
+
+        update_or_push(&mut state_list, state_index, Box::new(state));
+    } else if let ContextFor::SetState { set_state_item, .. } = &ctx.context_for {
+        let sig_id = set_state_item.sig_id();
+
+        if sig_id.component_id == instance.component_id
+            && sig_id.id_type == SigIdType::State
+            && sig_id.index == state_index
+        {
+            match set_state_item {
+                SetStateItem::Set { sig_id: _, value } => {
+                    let mut_state = state_list.get_mut(state_index).unwrap();
+                    let next_value = value.lock().unwrap().take().unwrap();
+                    *mut_state = next_value;
+                }
+                SetStateItem::Mutate { sig_id: _, mutate } => {
+                    let state = state_list.get_mut(state_index).unwrap();
+                    let mutate = mutate.lock().unwrap().take().unwrap();
+                    mutate(state.as_mut());
+                }
+            }
+        }
+    }
+
+    let state: &State = state_list[state_index]
+        .as_ref()
+        .as_any()
+        .downcast_ref()
+        .unwrap();
+
+    let state: &State = unsafe { std::mem::transmute(state) };
```

Comment:
Found memory vulnerability here.
If instance dead, this reference would be point very wrong memory.

Reply:
It's not problem for `rendering_tree`'s `attach_event` because `rendering_tree` will be re-render right after update of component tree.
But the problem is `async` code. currently `Sig` is `'static` so it could be copy into `async move` statement and it will point wrong memory if the component is dead, but the async function is working yet.

Reply:
To solve this problem, we need to bind `use_state`'s lifetime, to do that, we need to put ctx as parameter.